File: backend/src/utils/metrics.js
/**
 * metrics.js — Prometheus Metrics Registry
 *
 * Centralised prom-client setup. Import and use the exported helpers
 * to record metrics from anywhere in the codebase.
 *
 * Exposed at GET /metrics in standard Prometheus text format.
 */

'use strict';

const promClient = require('prom-client');

// ── Default process/Node.js metrics (memory, CPU, event loop lag, etc.) ──────
const register = new promClient.Registry();
promClient.collectDefaultMetrics({ register, prefix: 'nodejs_' });

// ─── HTTP Request Metrics ─────────────────────────────────────────────────────

const httpRequestsTotal = new promClient.Counter({
  name: 'http_requests_total',
  help: 'Total number of HTTP requests',
  labelNames: ['method', 'route', 'status_code'],
  registers: [register],
});

const httpRequestDurationSeconds = new promClient.Histogram({
  name: 'http_request_duration_seconds',
  help: 'Duration of HTTP requests in seconds',
  labelNames: ['method', 'route', 'status_code'],
  buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2, 5],
  registers: [register],
});

// ─── BullMQ Queue Metrics ─────────────────────────────────────────────────────

const bullmqQueueWaiting = new promClient.Gauge({
  name: 'bullmq_queue_waiting',
  help: 'Number of jobs waiting in a BullMQ queue',
  labelNames: ['queue'],
  registers: [register],
});

const bullmqQueueActive = new promClient.Gauge({
  name: 'bullmq_queue_active',
  help: 'Number of jobs actively processing in a BullMQ queue',
  labelNames: ['queue'],
  registers: [register],
});

const bullmqJobsTotal = new promClient.Counter({
  name: 'bullmq_jobs_total',
  help: 'Total number of BullMQ jobs processed',
  labelNames: ['queue', 'status'],  // status: completed | failed
  registers: [register],
});

const bullmqJobDurationSeconds = new promClient.Histogram({
  name: 'bullmq_job_duration_seconds',
  help: 'Duration of BullMQ job processing in seconds',
  labelNames: ['queue'],
  buckets: [0.1, 0.5, 1, 2, 5, 10, 30, 60],
  registers: [register],
});

// ─── Database Pool Metrics ────────────────────────────────────────────────────

const dbPoolTotal = new promClient.Gauge({
  name: 'pg_pool_total_connections',
  help: 'Total connections in the PostgreSQL pool',
  registers: [register],
});

const dbPoolIdle = new promClient.Gauge({
  name: 'pg_pool_idle_connections',
  help: 'Idle connections in the PostgreSQL pool',
  registers: [register],
});

const dbPoolWaiting = new promClient.Gauge({
  name: 'pg_pool_waiting_clients',
  help: 'Clients waiting for a connection from the pool',
  registers: [register],
});

// ─── GitHub API Metrics ───────────────────────────────────────────────────────

const githubApiCallsTotal = new promClient.Counter({
  name: 'github_api_calls_total',
  help: 'Total GitHub API calls made by the backend',
  labelNames: ['endpoint', 'status'],
  registers: [register],
});

const githubApiDurationSeconds = new promClient.Histogram({
  name: 'github_api_duration_seconds',
  help: 'Duration of GitHub API calls in seconds',
  labelNames: ['endpoint'],
  buckets: [0.1, 0.25, 0.5, 1, 2, 5],
  registers: [register],
});

// ─── Middleware: HTTP metrics ──────────────────────────────────────────────────

/**
 * Express middleware that records request count and duration for every response.
 * Normalises dynamic route segments (e.g. /repos/:id/scan → /repos/:id/scan)
 * using the matched Express route pattern.
 */
function httpMetricsMiddleware(req, res, next) {
  const startNs = process.hrtime.bigint();

  res.on('finish', () => {
    const durationSec = Number(process.hrtime.bigint() - startNs) / 1e9;
    // Use matched route pattern if available, else fall back to raw path
    const route      = req.route?.path || req.path || 'unknown';
    const routeKey   = `${req.method} ${route}`;
    const statusCode = String(res.statusCode);

    httpRequestsTotal.inc({ method: req.method, route: routeKey, status_code: statusCode });
    httpRequestDurationSeconds.observe({ method: req.method, route: routeKey, status_code: statusCode }, durationSec);
  });

  next();
}

// ─── Prometheus text-format handler ──────────────────────────────────────────

async function metricsHandler(req, res) {
  try {
    res.set('Content-Type', register.contentType);
    res.end(await register.metrics());
  } catch (err) {
    res.status(500).end(err.message);
  }
}

module.exports = {
  register,
  httpMetricsMiddleware,
  metricsHandler,
  // Exported so other modules can record metrics
  httpRequestsTotal,
  httpRequestDurationSeconds,
  bullmqQueueWaiting,
  bullmqQueueActive,
  bullmqJobsTotal,
  bullmqJobDurationSeconds,
  dbPoolTotal,
  dbPoolIdle,
  dbPoolWaiting,
  githubApiCallsTotal,
  githubApiDurationSeconds,
};

File: ai/utils/metrics.py
"""
metrics.py — Prometheus Metrics for DevPulse AI Service

Centralised prometheus-client registry.
Exposed at GET /metrics (Prometheus text format).
"""

from prometheus_client import (
    CollectorRegistry,
    Counter,
    Histogram,
    generate_latest,
)

# ─── Registry ─────────────────────────────────────────────────────────────────
# Use the default global registry (works fine for a single-process uvicorn service)
registry = CollectorRegistry(auto_describe=True)

# ─── AI Analysis Metrics ──────────────────────────────────────────────────────

ai_analysis_requests_total = Counter(
    "ai_analysis_requests_total",
    "Total number of AI analysis requests",
    ["status"],  # status: success | error | fallback
    registry=registry,
)

ai_analysis_duration_seconds = Histogram(
    "ai_analysis_duration_seconds",
    "Duration of AI analysis pipeline in seconds",
    buckets=[0.1, 0.25, 0.5, 1, 2, 5, 10, 20, 30, 60],
    registry=registry,
)

ai_llm_calls_total = Counter(
    "ai_llm_calls_total",
    "Total LLM API calls made by the AI service",
    ["model", "status"],  # status: success | error | ratelimited
    registry=registry,
)

ai_llm_duration_seconds = Histogram(
    "ai_llm_duration_seconds",
    "Duration of individual LLM API calls in seconds",
    ["model"],
    buckets=[0.5, 1, 2, 5, 10, 20, 30],
    registry=registry,
)

# ─── HTTP Metrics (supplementing OTel) ────────────────────────────────────────

http_requests_total = Counter(
    "http_requests_total",
    "Total HTTP requests handled by AI service",
    ["method", "route", "status_code"],
    registry=registry,
)

http_request_duration_seconds = Histogram(
    "http_request_duration_seconds",
    "HTTP request duration in seconds",
    ["method", "route", "status_code"],
    buckets=[0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2, 5],
    registry=registry,
)


def get_metrics_output() -> bytes:
    """Return Prometheus text format bytes for /metrics endpoint."""
    return generate_latest(registry)
